src/mlpack/methods/matrix_completion/mc_fw_update_matrix.hpp

```diff
@@ -0,0 +1,238 @@
+/**
+ * @file mc_fw_update_matrix.hpp
+ * @author Chenzhe Diao
+ *
+ * In FrankWolfe algorithm, used as UpdateRuleType.
+ * Designed for matrix completion problems.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_MATRIX_COMPLETION_MC_FW_UPDATE_MATRIX_HPP
+#define MLPACK_METHODS_MATRIX_COMPLETION_MC_FW_UPDATE_MATRIX_HPP
+
+#include "mc_fw_function.hpp"
+#include <mlpack/core/optimizers/proximal/proximal.hpp>
+
+namespace mlpack {
+namespace matrix_completion {
+
+/**
+ * Update rule for matrix completion problem using Frank-Wolfe type solver.
+ *
+ * The implementation follows:
+ * @code
+ * Rao N, Shah P, Wright S  (2015)
+ * Forward--backward greedy algorithms for atomic norm regularization.
+ * IEEE Transactions on Signal Processing 63(21):5798–5811
+ * @endcode
+ *
+ * Since the nuclear norm could be viewed as an atom norm, the code here also
+ * contains the atom operations, such as AddAtom() and RecoverMatrix().
+ *
+ * The implementation also added the enhancement step and truncation step to
+ * make the algorithm converge faster.
+ */
+class UpdateMatrix {
+ public:
+  UpdateMatrix(double tau): tau(tau)
+  {/* Nothing to do. */}
+
+  void Update(MatrixCompletionFWFunction& function,
+              const arma::mat& oldCoords,
+              const arma::mat& s,
+              arma::mat& newCoords,
+              const size_t numIter)
+  {
+    // Line search, with explicit solution here.
+    arma::mat v = tau * s - oldCoords;
+    arma::vec b = function.Values();
+
+    arma::vec oldEntries, vEntries;
+    function.GetKnownEntries(oldCoords, oldEntries);
+    function.GetKnownEntries(v, vEntries);
+
+    double gamma = arma::dot(b - oldEntries, vEntries);
```

Comment:
Where does this sequence of updates for gamma come from? If it suggested by a paper, can you leave a short comment with a pointer.